### ISSUE TITLE
JsonExtensionData related issues

### ISSUE THREAD

**Issue by stevecasl** [JamesNK/Newtonsoft.Json#852]
I've just been using this feature and I've found a heap of issues related to this, some tightly bound to the issue and some more wider reaching.

1/ Detection of default constructor.
ReflectionUtil::GetDefaultConstructor check is not comprehensive.
The check is that there exists a constructor with NO parameters.
This fails with
public MyClass(MyClass other = null) { ... }

Since reflection indicates it does indeed have a parameter, however it is optional and defaulted. My view is that this is a matching constructor.
Expanding, it should be a check for all of the constructor parameters being optional/defaulted.

This rolls onto issue 2

2/ Extension data ends up with all of the non-json class property names in it when attempting to create an instance.

[JsonObject(MemberSerialization.OptIn)]
class MyClass
{
public int NotForJson { get; set; }

[JsonPropertyAttribute(Required = Requred.Always)]
public int ForJson { get; set; }

[JsonExtensionData(ReadData = true, WriteData = true)]
public IDictionary<String, JToken> ExtraInfoJson { get; set;   }

public MyClass(MyClass other = null)
{
  if( other != null)
  {
      // copy construct
  }
}
}
jsonStirng = { "ForJson" : 33 , "extra1" : 11, "extra2" : 22 }

This results in ExtraInfoJson have 3 entries, the extra1 and extra2 but with a NotForJson with a null value.

I believe the issue is at JsonSerializationInternalReader::PopulateObject where one or more of the checks should see if the value of the data is null or not.

Less clear on if it's a bug, in this code if I had a value for an ignored member it would be added to the extension data. Semantically I see that as erroneous input JSON data, it should be stored in a different data entity, some kind of JsonIgnoredData type attributed member.

3/ Unit test Serialization.ExtensionDataTests::ExtensionDataTest appears wrong
I started by looking at the unit test and I found I couldn't understand some of the reasoning and that perhaps the test could be improved. The tests may highlight some more issues.

Assert for the ReadOnly item - should also check that the instantiated class readonly member hasn't been changed? as it stands the test has the same text making it unclear if it should or should not, and if actually does modify the member.
The assert for Ints list expects there to be 4 values. The Constructor for the class creates the list with one item, the deserialisation is appending the 3 JSON text values to this, thus the object created from the JSON != the JSON data. It should be 3 long and have 1,2,3 ? This looks like a clear bug.
GetPrivate member looks to have been initialised by the JSON to true, reading between the lines, since it appears to be expected in the extension data, I would it expect to have the default false value. I suspect a bug.

The assert for the extensionData count being 7 obviously depends on the intention behind "extension data". My interpretation is that it should contain the extra JSON data above the valid object data.
That makes the case for ignored parameters ambiguous, they perhaps should appear in the list, however because they will serialised out again, that seems to be a way to continue the passage of  unintended data, after all the ignored field value will appear to be present when written out, but is likely to not be the value of the in memory object.
Hence my feeling this might be classified as recognised but invalid.

These look like changes that might sweep quite far, so while I could submit a patch, I suspect someone who understands the issues would be better involved in it. My patch would address these issues but probably cause knock ons for users and other tests alike, making any patch get bigger and bigger. I suspect I'd be wasting my time as you'd then go and look at this issue.

It's a great library, I've used it a few times in the past, this is the first time I tried to use this feature though (and is exactly what I was after, grabbing the unused bits of JSON after initialising my object so that other derived classes could use it). I'm hoping I can work around these bugs over the next few days.

**Comment by JamesNK** [comment]
1. Treating non-default constructors as default is really complex. For now I want to keep it simple. Just put an attribute on the constructor.
2. Works for me:

        [JsonObject(MemberSerialization.OptIn)]
        public class MyClass
        {
            public int NotForJson { get; set; }

            [JsonPropertyAttribute(Required = Required.Always)]
            public int ForJson { get; set; }

            [JsonExtensionData(ReadData = true, WriteData = true)]
            public IDictionary<String, JToken> ExtraInfoJson { get; set; }

            public MyClass(MyClass other = null)
            {
                if (other != null)
                {
                    // copy construct
                }
            }
        }

        [Test]
        public void PopulateWithExtensionData()
        {
            string jsonStirng = @"{ ""ForJson"" : 33 , ""extra1"" : 11, ""extra2"" : 22 }";

            MyClass c = new MyClass();

            JsonConvert.PopulateObject(jsonStirng, c);

            Assert.AreEqual(2, c.ExtraInfoJson.Count);
            Assert.AreEqual(11, (int)c.ExtraInfoJson["extra1"]);
            Assert.AreEqual(22, (int)c.ExtraInfoJson["extra2"]);
        }

3.
You can't set readonly, therefor it is extension data.
Arrays are appended to by default. You can modify this with settings.
You can't set an internal property to GetPrivate is extension data.

tl;dr; extension data is anything that can't be set during deserialization

**Comment by stevecasl** [comment]
I wonder if your test passed because you used PopulateObject rather than DeserializeObject<MyClass>(...) which actually searches for and invokes the constructor.
That would also explain the arrays appended to.
I'd still hold that if the JSON data passed through deserialise object doesn't match the actual class data for serialised members it is a failure, it fails the least surprise test for me.

**Comment by ravensorb** [comment]
I agree. Also I too have noticed the DeserializeObject behaves differently than PopulateObject
